gc.c

```diff
@@ -5045,6 +5083,10 @@ static void invalidate_moved_page(rb_objspace_t *objspace, struct heap_page *pag
 static void
 read_barrier_handler(uintptr_t address)
 {
+    if (!is_pointer_to_heap_body(&rb_objspace, address)) {
+        rb_bug("read barrier activated for non-heap memory\n");
```

Comment:
This obfuscates the error original error message as well. Instead of "couldn't unprotect memory", you get "read barrier activated for non-heap memory", neither of which is the original error. If we want to get the original message we should call the signal handler for that signal. But I think that would be confusing too, because the stack trace would refer to this point as the source of the crash (when it really is further up the the stack).

Reply:
The segv handler returns the address exactly where the fault occurs - which is not necessarily aligned to an RVALUE. `is_pointer_to_heap` relies on the address being at the start of an RVALUE, otherwise it returns false.